backend/data/src/main/kotlin/io/tolgee/batch/BatchJobProjectLockingManager.kt

```diff
@@ -71,65 +77,84 @@ class BatchJobProjectLockingManager(
+    return computedJobIds.contains(batchJobDto.id)
   }
 
-  fun getLockedForProject(projectId: Long): Long? {
+  fun getLockedForProject(projectId: Long): Set<Long> {
     if (usingRedisProvider.areWeUsingRedis) {
-      return getRedissonProjectLocks()[projectId]
+      return getRedissonProjectLocks()[projectId] ?: emptySet()
     }
-    return localProjectLocks[projectId]
+    return localProjectLocks[projectId] ?: emptySet()
   }
 
-  private fun tryLockLocal(toLock: BatchJobDto): Boolean {
-    val projectId = toLock.projectId ?: return true
-    val computed =
-      localProjectLocks.compute(projectId) { _, value ->
-        val newLocked = computeFnBody(toLock, value)
-        logger.debug("While trying to lock ${toLock.id} for project ${toLock.projectId} new lock value is $newLocked")
-        newLocked
-      }
-    return computed == toLock.id
+  private fun tryLockLocal(batchJobDto: BatchJobDto): Boolean {
+    val projectId = batchJobDto.projectId ?: return true
+    val computedJobIds =
+      localProjectLocks.compute(projectId) { _, lockedJobIds ->
+        val newLockedJobIds = computeFnBody(batchJobDto, lockedJobIds ?: emptySet())
+        logger.debug(
+          "While trying to lock locally {} for project {} new lock value is {}",
+          batchJobDto.id,
+          batchJobDto.projectId,
+          newLockedJobIds
+        )
+        newLockedJobIds
+      } ?: emptySet()
+    return computedJobIds.contains(batchJobDto.id)
   }
 
   private fun computeFnBody(
     toLock: BatchJobDto,
-    currentValue: Long?,
-  ): Long {
+    lockedJobIds: Set<Long>,
+  ): Set<Long> {
     val projectId =
       toLock.projectId
         ?: throw IllegalStateException(
           "Project id is required. " +
             "Locking for project should not happen for non-project jobs.",
         )
-    // nothing is locked
-    if (currentValue == 0L) {
-      logger.debug("Locking job ${toLock.id} for project ${toLock.projectId}, nothing is locked")
-      return toLock.id
-    }
 
-    // value for the project is not initialized yet
-    if (currentValue == null) {
+    // nothing is locked
+    if (lockedJobIds.isEmpty()) {
       logger.debug("Getting initial locked state from DB state")
       // we have to find out from database if there is any running job for the project
-      val initial = getInitialJobId(projectId)
-      logger.debug("Initial locked job $initial for project ${toLock.projectId}")
-      if (initial == null) {
-        logger.debug("No job found, locking ${toLock.id}")
-        return toLock.id
+      val initialJobId = getInitialJobId(projectId)
+      logger.info("Initial locked job $initialJobId for project ${toLock.projectId}")
+      if (initialJobId == null) {
+        logger.debug("No initial job found, locking only ${toLock.id}")
+        return setOf(toLock.id)
       }
+      val newLockedJobIds = mutableSetOf<Long>(initialJobId)
+      if (newLockedJobIds.size < batchProperties.projectConcurrency) {
```

Comment:
won't `newLockedJobIds.size` be always equal to 1 here?

Reply:
It seemed I handled the case where `projectConcurrency` could be 0...
Or prepared for `getInitialJobId` to return multiple ids... 😄 

YAGNI, will remove